src/gobupload/storage/update_table.py

```diff
@@ -0,0 +1,612 @@
+    FROM (
+        SELECT * FROM {self.src_table_name} WHERE {FIELD.DATE_DELETED} IS NULL
+    ) src
+    {validgeo_src}
+    {self._join_array_elements() if self.is_many else ""}
+    LEFT JOIN {dst} dst ON {self.and_join.join(self._dst_table_inner_join_on())}
+    GROUP BY {self.comma_join.join(self._src_dst_group_by())}
+) src_dst ON {self.and_join.join(self._src_dst_join_on())}
+"""
+
+    def _get_query(self):
+        """Builds and returns the event extraction query
+
+        :return:
+        """
+
+        rel_table_join_on = self._rel_table_join_on()
+        dst_table_outer_join_on = self._dst_table_outer_join_on()
+        select_expressions = self._select_expressions()
+
+        joins = f"""
+{self._join_array_elements() if self.is_many else ""}
+FULL JOIN {self.relation_table} rel
+    ON {self.and_join.join(rel_table_join_on)}
+{self._src_dst_join()}
+LEFT JOIN {self.dst_table_name} dst
+    ON {self.and_join.join(dst_table_outer_join_on)}
+"""
+
+        return f"""
+SELECT
+    {self.comma_join.join(select_expressions)}
+FROM (
+    SELECT * FROM {self.src_table_name} WHERE {FIELD.DATE_DELETED} IS NULL
+) src
+{joins}
+{self._where()}
+"""
+
+    def extract(self):
+        """Extracts/generates events from the current state of the database.
+
+        The current results are close to GOB events, but the real implementation of GOB events should be done later on.
+
+        :return:
+        """
+        query = self._get_query()
+
+        return _execute(query)
+
+
+class RelationTableUpdater:
+    """RelationTableUpder updates the relation table based on the observed relations in the model.
+
+    This class is written with the future use of real GOB events in mind.
+
+    Usage:
+
+    updater = RelationTableUpdater(catalog, collection, relation_field_name)
+    update_cnt = updater.update_relation()
+
+    """
+    MAX_QUEUE_LENGTH = 1000
+
+    def __init__(self, src_catalog_name, src_collection_name, src_field_name):
+        self.src_catalog_name = src_catalog_name
+        self.src_collection_name = src_collection_name
+        self.src_field_name = src_field_name
+        self.events_extractor = RelationTableEventExtractor(self.src_catalog_name, self.src_collection_name,
+                                                            self.src_field_name)
+        self.relation_table = self.events_extractor.relation_table
+        self.fields = self._get_fields()
+        self.update_cnt = 0
+
+    def _get_fields(self):
+        """Returns the field mapping of database fields to the fields in the events as received from the
+        EventExtractor.
+
+        :return:
+        """
+        fields = {
+            FIELD.GOBID: 'rel__gobid',
+            FIELD.ID: None,
+            FIELD.SOURCE: 'src__source',
+            FIELD.APPLICATION: 'src__application',
+            FIELD.SOURCE_ID: 'src__source_id',
+            FIELD.VERSION: 'src__version',
+            'id': None,
```

Comment:
De relatie tabel heeft denk ik ook een eigen versioning. Want als het formaat daarvan wijzigt dan is dat een nieuwe versie van de relatie tabel. Dus version is denk ik niet de src__version maar gewoon 0.1